lib/cstruct.ml

```diff
@@ -254,41 +255,869 @@ external swap64 : int64 -> int64 = "%bswap_int64"
+(* Here we also assume there is no trailing padding at the end *)
+(* The offset and alignment rules depend on the "mother" type *)
+let align = 8 (* bits, not bytes *)
+
+(* These functions do not cross the boundary, and just add zeroes *)
+let get_uint16_chop_boundary swap p t i =
+    if i + 2 > t.len then
+        let part = get_uint8 t i in
+        if swap then part lsl 8 else part
+    else
+        get_uint16 swap p t i
+
+let get_uint32_chop_boundary swap p t i =
+    if i + 4 > t.len then
+        if i + 3 > t.len then
+            Int32.of_int (get_uint16_chop_boundary swap p t (i + 2))
+        else
+            let part1 = get_uint16 swap p t i in
+            let part2 = get_uint8 t (i + 2) in
+            let chunk =
+                if swap then part1 lsl 8 + part2 else part2 lsl 16 + part1
+            in
+            Int32.of_int (chunk lsl 8)
+    else
+        get_uint32 swap p t i
+
+let get_uint64_chop_boundary swap p t i =
+    if i + 8 > t.len then
+        if i + 7 > t.len then
+            if i + 6 > t.len then
+                if i + 5 > t.len then
+                    Int64.of_int32 (get_uint32_chop_boundary swap p t (i + 4))
+                else
+                    let part1 = Int64.of_int32 (get_uint32 swap p t i) in
+                    let part2 = Int64.of_int (get_uint8 t (i + 4)) in
+                    let chunk =
+                        if swap then
+                            Int64.add (Int64.shift_left part1 8) part2
+                        else
+                            Int64.add (Int64.shift_left part2 32) part1
+                    in
+                    Int64.shift_left chunk 24
+            else
+                let part1 = Int64.of_int32 (get_uint32 swap p t i) in
+                let part2 = Int64.of_int (get_uint16 swap p t (i + 4)) in
+                let chunk =
+                    if swap then
+                        Int64.add (Int64.shift_left part1 16) part2
+                    else
+                        Int64.add (Int64.shift_left part2 32) part1
+                in
+                Int64.shift_left chunk 16
+        else
+            let part1 = Int64.of_int32 (get_uint32 swap p t i) in
+            let part2 = Int64.of_int32 (get_uint32_chop_boundary swap p t (i + 4)) in
+            let chunk =
+                if swap then
+                    Int64.add (Int64.shift_left part1 32) part2
+                else
+                    Int64.add (Int64.shift_left part2 32) part1
+            in
+            chunk
+    else
+        get_uint64 swap p t i
+
+
+(* It also should always start from the most significant bit for big endian and least significant bit on little endian machine
+ * Basically on big endian machine we should read them in reversed order, bit by bit *)
+(* Currently it always starts from the least significant bit
+ * How to handle the byte-peeking then? *)
+(* This function cannot read more than 8 bits *)
+let get_bits_uint8_le swap p t o i =
```

Comment:
Hmmhmm, it seems that the predicate on this function is to be able to access at, at most, 2 bytes when an asked series of bits can overlap 2 bytes (like, give me 8 bits when `off = 4`). I think, this predicate adds a big complexity when the _at most_ can not be assert at the beginning - and this why, if I understand correctly, we need `get_uint16_chop_boundary`.

I think we can split this function (for instance, I'm focus on this function), to little functions here. A safe one which did not need `get_uint16_chop_boundary` (easy case) and the other complex case. It's much more about readability.